src/main/cpp/util/InstallDirs.cpp

```diff
@@ -170,19 +171,36 @@ path InstallDirs::GetRootDir()
 	return g_root_dir;
 }
 
+std::unique_ptr<boost::filesystem::ifstream> InstallDirs::OpenFile(
+        const boost::filesystem::path& relative_path,
+        const boost::filesystem::path& anchor_path)
+{
+	auto file_path = anchor_path / relative_path;
+	if (!is_regular_file(file_path)) {
+		FATAL_ERROR("File " + file_path.string() + " not present.");
+	}
+
+        auto stream = std::make_unique<boost::filesystem::ifstream>();
+	stream->open(file_path.string());
+	if (!stream->is_open()) {
+		FATAL_ERROR("Error opening file " + file_path.string());
+	}
+
+        return stream;
+}
+
 void InstallDirs::ReadXmlFile(
 	const boost::filesystem::path& relative_path,
 	const boost::filesystem::path& anchor_path,
 	boost::property_tree::ptree& result)
 {
-	auto file_path = InstallDirs::GetDataDir() /= relative_path;
-	if (!is_regular_file(file_path)) {
-		FATAL_ERROR(
-			"File " + relative_path.string() +
-			" not present. (full path: " + file_path.string() + ")");
-	}
+	boost::property_tree::read_xml(*OpenFile(relative_path, anchor_path), result);
```

Comment:
`OpenFile` returns an `std::unique_ptr<boost::filesystem::ifstream>` right? I'm not sure if it's save to dereference an `std::unique_ptr<boost::filesystem::ifstream>`. I have a hunch that `OpenFile(...)`'s lifetime ends when the expression it is used in (i.e., the pointer dereference, *not* the call to `read_xml`) ends. If I'm right, then `read_xml` is accessing a pointer to memory that's already been deallocated, which is UB.